pennylane/templates/state_preparations/arbitrary_state_preparation.py

```diff
@@ -79,19 +62,53 @@ def ArbitraryStatePreparation(weights, wires):
 
         @qml.qnode(dev)
         def vqe(weights):
-            qml.ArbitraryStatePreparations(weights, wires=[0, 1, 2, 3])
+            qml.templates.ArbitraryStatePreparation(weights, wires=[0, 1, 2, 3])
 
             return qml.expval(qml.Hermitian(H, wires=[0, 1, 2, 3]))
 
+    The shape of the weights parameter can be computed as follows:
+
+    .. code-block:: python
+
+        shape = qml.templates.ArbitraryStatePreparation.shape(n_wires=3)
+
+
     Args:
-        weights (tensor_like): The angles of the Pauli word rotations, needs to have length :math:`2^(n+1) - 2`
+        weights (tensor_like): Angles of the Pauli word rotations. Needs to have length :math:`2^(n+1) - 2`
             where :math:`n` is the number of wires the template acts upon.
-        wires (Iterable or Wires): Wires that the template acts on. Accepts an iterable of numbers or strings, or
-            a Wires object.
+        wires (Iterable): wires that the template acts on
     """
 
-    wires = Wires(wires)
-    _preprocess(weights, wires)
+    num_params = 1
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(self, weights, wires, do_queue=True):
+
+        shape = qml.math.shape(weights)
+        if shape != (2 ** (len(wires) + 1) - 2,):
+            raise ValueError(
+                f"Weights tensor must be of shape {(2 ** (len(wires) + 1) - 2,)}; got {shape}."
+            )
+
+        super().__init__(weights, wires=wires, do_queue=do_queue)
+
+    def expand(self):
+
+        with qml.tape.QuantumTape() as tape:
+            for i, pauli_word in enumerate(_state_preparation_pauli_words(len(self.wires))):
```

Comment:
There is actually a reason I am trying to use indices: Indexing always works nicely through all interfaces. But iterating over tensors does not, for example there were issues with tf tensors that are not iterables. Also I suspect that indexing is faster.

I have to admit it would be good to systematically check this, but for now I am trying to stick to iterating over indices...

Reply:
Thanks for the explanation :)

So I just ran a quick check, and it seems like using zip is *slightly* faster. but it's more important to make sure it works in all interfaces.